theano/gpuarray/opt.py

```diff
@@ -215,6 +256,242 @@ def apply(self, fgraph):
                     0, 'fast_run', 'fast_compile', 'merge')
 
 
+class GraphToGPU(NavigatorOptimizer):
+    """
+    Transfer the graph as a whole to GPU instead of transfering node by node.
+    """
+
+    def __init__(self, local_optimizers_all, local_optimizers_map):
+        self.local_optimizers_all = local_optimizers_all
+        self.local_optimizers_map = local_optimizers_map
+        self.failure_callback = None
+
+    def add_requirements(self, fgraph):
+        fgraph.attach_feature(toolbox.ReplaceValidate())
+
+    def apply(self, fgraph):
+        mapping = {}
+        time_opts = {}
+        node_created = {}
+        process_count = {}
+        t_topo = time.time()
+        topo = fgraph.toposort()
+        time_topo = time.time()
+        toposort_timing = time_topo - t_topo
+
+        # Building a new graph
+        # Iterating through inputs of graph
+        target = infer_context_name(*fgraph.inputs)
+        for i in fgraph.inputs:
+            # Do not move *int* scalar to the GPU.
+            target = getattr(i.tag, 'target', None)
+            if (target != 'cpu' and
+               isinstance(i.type, tensor.TensorType) and
+               (i.ndim > 0 or 'int' not in i.dtype)):
+                mapping[i] = as_gpuarray_variable(i, target)
```

Comment:
@abergeron the way we support target here is mostly the same as in the class InputToGraphOptimization. So if there is a problem here, it came from there.

I think as_gpuarray_variable should handle the possible bad target, not here. We don't want to handle bad target at many places, just one or a few and as_gpuarray_variable seem a good place for that. If it don't do that, it is for another PR.

@abergeron the condition is so that all non-int are on the GPU and all ndim > 0 are on GPU. So mostly, we only prevent scalar int from going to the GPU.